libraries/chain/fork_database.cpp

```diff
@@ -571,28 +680,75 @@ namespace eosio::chain {
                   "block state not in fork database; cannot mark as valid",
                   ("id", h->id()) );
 
-      by_id_idx.modify( itr, []( bsp& _bsp ) {
-         _bsp->set_valid(true);
+      by_id_idx.modify( itr, []( auto& i ) {
+         BSAccessor::set_valid(*i, true);
       } );
 
-      auto candidate = index.template get<by_lib_block_num>().begin();
+      auto candidate = index.template get<by_best_branch>().begin();
       if( first_preferred( **candidate, *head ) ) {
          head = *candidate;
       }
    }
 
-   template<class bsp>
-   bsp fork_database_t<bsp>::get_block(const block_id_type& id) const {
+   template<class BSP>
+   void fork_database_t<BSP>::update_best_qc( const block_id_type& id, const qc_claim_t& best_qc_claim ) {
+      std::lock_guard g( my->mtx );
+      my->update_best_qc_impl( id, best_qc_claim );
+   }
+
+   template<class BSP>
+   void fork_database_impl<BSP>::update_best_qc_impl( const block_id_type& id, const qc_claim_t& best_qc_claim ) {
+      auto& by_id_idx = index.template get<by_block_id>();
+
+      auto itr = by_id_idx.find( id );
+      EOS_ASSERT( itr != by_id_idx.end(), fork_database_exception,
+                  "block state not in fork database; cannot update", ("id", id) );
+
+      if (BSAccessor::qc_claim_update_needed(**itr, best_qc_claim)) {
+         by_id_idx.modify( itr, [&]( auto& i ) {
+            BSAccessor::update_best_qc(*i, best_qc_claim);
+         } );
+      }
+
+      // process descendants
+      vector<block_id_type> descendants;
+      descendants.reserve(index.size());
+      descendants.push_back(id);
+      auto& previdx = index.template get<by_prev>();
+      for( uint32_t i = 0; i < descendants.size(); ++i ) {
+         auto previtr = previdx.lower_bound( descendants[i] );
+         while( previtr != previdx.end() && (*previtr)->previous() == descendants[i] ) {
+            if (BSAccessor::qc_claim_update_needed(**previtr, best_qc_claim)) {
+               previdx.modify( previtr, [&](auto& i) {
+                  BSAccessor::update_best_qc(*i, best_qc_claim);
+               });
+            } else {
```

Comment:
Doesn't this while loop iterate over all the immediate children of `descendants[i]`?

If that's the case, we don't want to `break;` and not visit the siblings. We want to not add this child to the `descendants` vector, so we won't consider its children.

Reply:
I think you are right. Good catch. Seems I should add a unittest for this.

Reply:
Added a test that failed before fixing the logic.